storage-provider/server/src/retrieval/blockstore.rs

```
@@ -0,0 +1,206 @@
+        let mut raw_piece = File::open(&raw_pieces_path).await.map_err(|_| {
+            blockstore::Error::StoredDataError(format!("piece {:?} not found", raw_pieces_path))
+        })?;
+        raw_piece
+            .seek(SeekFrom::Start(block.offset))
+            .await
+            .map_err(|_| {
+                blockstore::Error::StoredDataError(
+                    "error occurred while seeking the raw piece".into(),
+                )
+            })?;
+
+        // Read block data
+        let mut buffer = vec![0; block.size as usize];
+        raw_piece.read_exact(&mut buffer).await.map_err(|err| {
+            blockstore::Error::StoredDataError(format!(
+                "error occurred while reading the raw piece: {}",
+                err.to_string()
+            ))
+        })?;
+
+        Ok(Some(buffer))
+    }
+
+    async fn put_keyed<const S: usize>(
+        &self,
+        _cid: &cid::CidGeneric<S>,
+        _data: &[u8],
+    ) -> blockstore::Result<()> {
+        Err(Error::FatalDatabaseError(
+            "put operation not supported".to_string(),
+        ))
+    }
+
+    async fn remove<const S: usize>(&self, _cid: &cid::CidGeneric<S>) -> blockstore::Result<()> {
+        Err(Error::FatalDatabaseError(
+            "remove operation not supported".to_string(),
+        ))
+    }
+
+    async fn close(self) -> blockstore::Result<()> {
+        Ok(())
+    }
+}
+
+/// Convert CID with the generic Multihash size to the CID with the specific
+/// Multihash size that the underlying blockstore expects.
+fn to_blockstore_cid<const S: usize>(cid: &CidGeneric<S>) -> Result<Cid, Error> {
+    let digest_size = cid.hash().size() as usize;
+    let hash = cid
+        .hash()
+        .resize::<64>()
+        .map_err(|_| Error::CidError(CidError::InvalidMultihashLength(digest_size)))?;
+
+    Ok(Cid::new(cid.version(), cid.codec(), hash).expect("we know cid is correct here"))
```

```suggestion
    Ok(Cid::new(cid.version(), cid.codec(), hash).expect("new CID is derived from a valid existing CID, this should never fail"))
```

---

Though we could also just return the error since we know it won't happen (?)

---

I've left an explicit indication of not failing